### PR TITLE
Process encrypted secret by the target

### LINKED ISSUE #7157: Process encrypted secret by the target
The target should check if there is the secret encrypted with its public key in the metadata of the locked transfer. 
If so the secret can be revealed to the previous hop. 
If not, the traditional way of requesting the secret continues.

### PR DESCRIPTION
**Pull request by netcriptus**
## Process encrypted secret by the target

Fixes: #7157

### CODE REVIEW

File: raiden/transfer/channel.py [hunk 1]
```diff
@@ -2548,68 +2548,85 @@ def state_transition(
...
-            withdraw=state_change,
-            block_number=block_number,
-            block_hash=block_hash,
-        )
-    elif type(state_change) == ReceiveWithdrawExpired:
-        assert isinstance(state_change, ReceiveWithdrawExpired), MYPY_ANNOTATION
-        iteration = handle_receive_withdraw_expired(
-            channel_state=channel_state, withdraw_expired=state_change, block_number=block_number
-        )
+    transition_map: Dict[Any, FuncMap] = {
```
Comment (ulope) [inline raiden/transfer/channel.py]: If you are refactoring this anyway I'd suggest to convert all the `handle_*` functions into generic singledispatch functions (using `functools.singledispatch`) instead of a homegrown solution.

File: raiden/message_handler.py [hunk 2]
```diff
@@ -338,6 +339,13 @@ def handle_message_lockedtransfer(
         sender = from_transfer.balance_proof.sender
 
         if message.target == TargetAddress(raiden.address):
+            encrypted_secret = getattr(message.metadata, "encrypt_secret", None)
```
Comment (ulope) [inline raiden/message_handler.py]: Is `getattr` needed?
How can the attribute ever be missing?

File: raiden/message_handler.py [hunk 3]
```diff
@@ -338,6 +339,13 @@ def handle_message_lockedtransfer(
...
+            if encrypted_secret:
```
Comment (ulope) [inline raiden/message_handler.py]: Better be explicit and use `is not None`.

File: raiden/message_handler.py [hunk 4]
```diff
@@ -338,6 +339,13 @@ def handle_message_lockedtransfer(
...
+                try:
+                    secret = decrypt_secret(encrypted_secret, raiden.rpc_client.privkey)
+                    return [ReceiveSecretReveal(secret=secret, sender=message.sender)]
+                except InvalidSecret:
+                    log.error("Ignoring invalid encrypted secret")
```
Comment (ulope) [inline raiden/message_handler.py]: It would be good to also log the sender address. 

The question is we should drop the transfer in this case since there's obviously something weird going on.

File: raiden/message_handler.py [hunk 5]
```diff
@@ -338,6 +339,13 @@ def handle_message_lockedtransfer(
         sender = from_transfer.balance_proof.sender
 
         if message.target == TargetAddress(raiden.address):
+            encrypted_secret = getattr(message.metadata, "encrypt_secret", None)
+            if encrypted_secret:
+                try:
+                    secret = decrypt_secret(encrypted_secret, raiden.rpc_client.privkey)
+                    return [ReceiveSecretReveal(secret=secret, sender=message.sender)]
```
Comment (ulope) [inline raiden/message_handler.py]: It would be nice to log that we're using the encrypted secret.

File: raiden/message_handler.py (hunk 4 above)
Comment (netcriptus) [reply raiden/message_handler.py]: It might be worth, at least on a first run, to fall back to request the secret if the encrypted secret received was invalid. Considering that this failure is logged, I would still think it's worth to keep the fallback behaviour on the first version of the feature.

File: raiden/message_handler.py [hunk 6]
```diff
@@ -338,6 +339,14 @@ def handle_message_lockedtransfer(
         sender = from_transfer.balance_proof.sender
 
         if message.target == TargetAddress(raiden.address):
+            encrypted_secret = message.metadata.encrypted_secret
+            if encrypted_secret is not None:
+                try:
+                    secret = decrypt_secret(encrypted_secret, raiden.rpc_client.privkey)
+                    log.info(f"Using encrypted secret received from {sender.hex()}")
```
Comment (fredo) [inline raiden/message_handler.py]: to encode addresses we should use `to_checksum_address()`

File: raiden/message_handler.py [hunk 7]
```diff
@@ -338,6 +339,14 @@ def handle_message_lockedtransfer(
...
+                    return [ReceiveSecretReveal(secret=secret, sender=message.sender)]
+                except InvalidSecret:
+                    log.error(f"Ignoring invalid encrypted secret received from {sender.hex()}")
```
Comment (fredo) [inline raiden/message_handler.py]: see above

**Review by ulope — APPROVED** [review]
LGTM

Very nice that we could get rid of the elif dispatch boilerplate...

**Review by fredo — DISMISSED** [review]
Im just requesting changes so that it does not get merged accidentally. We figured out that decryption of the secret is currently not tested and that it should not work because of a missing targetstate because `ActionTargetInit` is never applied.

File: raiden/message_handler.py (hunk 6 above)
Comment (fredo) [reply raiden/message_handler.py]: For debugging purposes it would be nice to have a constant event and things like sender can be attached with keywords.

File: raiden/tests/integration/api/rest/test_rest.py [hunk 8]
```diff
@@ -937,7 +939,10 @@ def test_channel_events_raiden(
 @pytest.mark.parametrize("number_of_nodes", [3])
 @pytest.mark.parametrize("channels_per_node", [CHAIN])
 @pytest.mark.parametrize("enable_rest_api", [True])
-def test_pending_transfers_endpoint(raiden_network: List[RaidenService], token_addresses):
+@patch("raiden.message_handler.decrypt_secret", side_effect=InvalidSecret)
```
Comment (ulope) [inline raiden/tests/integration/api/rest/test_rest.py:942]: We usually use pytest's `monkeypatch` fixture instead of `patch` in the test suite (although not always...).
The advantage is that the monkeypatch fixture is integrated with pytest and also defaults to raising an exception if the patch target doesn't exist (e.g. after a refactoring) which is not the case with `patch`.

I think it would be nicer to put this into a fixture and / or test helper (e.g. `disable_encrypted_secret`) so as not having to repeat this patch boilerplate on many tests (which would need to be changed if ever the `decrypt_secret` function was moved or otherwise refactored).

Also this only tests the branch where the included secret is invalid, not the one where none is provided in the first place.

**Review by ulope — APPROVED** [review]
This time LGTM for real ;)

Please see my comment on the test patching below.
This can be done in a follow up PR though so as not to hold this up any longer.

File: raiden/tests/integration/api/rest/test_rest.py (hunk 8 above)
Comment (netcriptus) [reply raiden/tests/integration/api/rest/test_rest.py:942]: Created https://github.com/raiden-network/raiden/issues/7287 for this
